Proposal for KRaft Support in Kafka Roller (#91)
strimzi/proposals: 060-kafka-roller-kraft.md

those checks and in what order but does not discuss exactly how the KafkaRoller works.
This proposal is expected to apply to both the current KafkaRoller and a future iteration of the KafkaRoller.

This proposal assumes that liveness/readiness of nodes is as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md) 
and [PR #8892](https://github.com/strimzi/strimzi-kafka-operator/pull/8892).

## Current situation

When operating on a ZooKeeper based cluster the KafkaRoller has the behaviour 
described below.

### Order to restart nodes
Restart in order:
 1. Unready brokers
 2. Ready brokers
 3. Current controller

### Triggers
The following are some of the triggers that restart a broker:
- Pod or its StrimziPodSet is annotated for manual rolling update
- Pod is unready
- Broker's configuration has changed

### Restart conditions
 1. Does not restart a broker performing log recovery
 2. Attempts to connect an admin client to a broker and if it can't connect, restarts the broker
 3. Restarts a broker if the pod is in stuck state
 4. Restarts a broker if the configuration has changed and cannot be updated dynamically

>  5. Does not restart a broker if doing so would take the in-sync replicas count below `min.insync.replicas` for any topic hosted on that broker

Author: KafkaRoller checks these conditions in order. Instead of splitting out into 2 lists, we added a sentence to clarify that the restart conditions are checked in the specified order.

#### Unready pod

> If pod is unready, KafkaRoller awaits its readiness before doing anything. If it hasn't become ready still, it checks whether it is stuck. A pod is considered stuck if it is in one of following states:

Author: I updated it to mention that the operational timeout is used for awaiting readiness.

> If pod is unready, KafkaRoller awaits its readiness before doing anything. If it hasn't become ready still, it checks whether it is stuck. A pod is considered stuck if it is in one of following states:

Reviewer: Operational timeout is 5 minutes -> that is way too long for stuck pods.

> If pod is unready, KafkaRoller awaits its readiness before doing anything. If it hasn't become ready still, it checks whether it is stuck. A pod is considered stuck if it is in one of following states:

Author: Checking the current design again, the statement about waiting before checking if pod is stuck is actually incorrect. We only wait for pod readiness if pod is not stuck. I will correct the statement.

- `CrashLoopBackOff`
- `ImagePullBackOff`
- `ContainerCreating`
- `Pending` and `Unschedulable`

If pod is unready and stuck, KafkaRoller restarts it if pod is out of date. Otherwise, KafkaRoller fails the reconcilation and does not restart other pods. This is because a pod being stuck could mean restarting the other pods will result in then being in the same state.

> If pod is unready but not stuck, KafkaRoller considers the restart conditions and may restart it.

Author: It now refers to the specific restart conditions it considers.

#### Configuration changes

>  - Retrieves the current Kafka configurations of the broker via admin client and compares it with the desired configurations specified in the Kafka CR. 

Author: Added a context introducing the list.

## Proposal

The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.

The behaviour when operating against a KRaft cluster is described below.

### Order to restart nodes
Restart in order:
1. Unready controller/combined nodes
2. Ready controller/combined nodes in follower state
3. Active controller (applies to both pure controller and combined node)
4. Unready broker-only nodes
5. Ready broker-only nodes

### Triggers
The following are some of the triggers that would restart a KRaft controller or combined node:
- Pod or its StrimziPodSet is annotated for manual rolling update
- Pod is unready 
- Controller's configuration has changed

The triggers for broker remain the same as Zookeeper mode.

### Restart conditions
The restart conditions that the KafkaRoller considers in different modes are described below.

#### The new quorum check
The restart conditions include a new check for controllers to verify that restarting the Kafka node does not affect the quorum health.

> The proposed check is:

Reviewer 2: ```suggestion
The proposed check ensures that a sufficient majority of controller nodes are caught up before allowing a restart:
```

- Check the quorum leader id using the quorum info and identify the `lastCaughtUpTimestamp` of the quorum leader
- Retrieve value of the Kafka property `controller.quorum.fetch.timeout.ms` from the desired configurations specified in the Kafka CR. If this property does not exist in the desired configurations, then use the hard-coded default value for it which is `2000`. The reason for this is explained further in the **NOTE** below.
- Mark a KRaft controller node as caught up if `leaderLastCaughtUpTimestamp - replicaLastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`, or if it is the current quorum leader. This will exclude the controller KafkaRoller is currently considering to restart, because we are trying to check whether the quorum would stay healthy when this controller is restarted.
- Count each controller node that is caught up (`numOfCaughtUpControllers`).
- Can restart if: `numOfCaughtUpControllers >= ceil((double) (totalNumOfControllers + 1) / 2)`.

> > NOTE: Until [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum) is implemented, KafkaRoller cannot create an admin connection to the controller directly to describe its configuration or the quorum state. Therefore KafkaRoller checks the desired configurations to get the value of `controller.quorum.fetch.timeout.ms` and creates admin client connection to the brokers for the quorum check. If KafkaRoller cannot connect to any of the brokers after the maximum attempts to retry, the controllers will be marked as failed to reconcile because of not being able to determine the quorum health. KafkaRoller would then try to reconcile the brokers, which may help restoring admin client connections to them. In this scenario, the reconcilation will be completed with failure and reported to the operator. The controllers quorum check will be retried in the next round of reconcilation.

Reviewer 2: I get page not found on the link

#### Separate brokers and controllers
For controller-only:
 1. Does not restart a controller performing log recovery
 2. Restarts a controller if the pod is in stuck state

>  3. Attempts to connect an admin client to any of the brokers, if cannot connect, return `UnforceableProblem` which results in delay and retry for the pod until the maximum attempts is reached. 

Reviewer 2: ```suggestion
 3. Attempts to connect an admin client to any of the brokers. If it cannot connect, returns `UnforceableProblem` which results in delay and retry for the pod until the maximum attempt is reached. 
```

When maximum attempts is reached the KafkaRoller will move on to restart broker pods incase this resolves connection issues but will still ultimately mark the reconciliation as failed.
 4. Restarts a controller if the controller configuration has changed
 5. Does not restart a controller if doing so would take the number of caught-up controllers (inc leader) to less than majority of the quorum

For broker-only:
 1. Does not restart a broker performing log recovery
 2. Restarts a broker if the pod is in stuck state
 3. Attempts to connect an admin client to a broker and if it can't connect restarts the broker

>  > **NEW** Currently when pod is stuck and out of date, it gets restarted because admin client connection fails. With this change, we will not attempt admin client connection to the broker, if the pod is stuck. Instead the pod will be restarted without making the admin client connection. Therefore the resulting action that restarts the pod that is stuck stays the same for zookeeper mode. This resolves the issue mentioned at the beginning of the proposal that leaves some pods stuck in pending state.

Reviewer 2: ```suggestion
 > **NEW** Currently when pod is stuck and out of date, it is restarted because admin client connection fails. With this change, we will not attempt admin client connection to the broker if the pod is stuck. Instead the pod will be restarted without making the admin client connection. Therefore, the resulting action that restarts the pod that is stuck stays the same for ZooKeeper mode. This resolves the issue mentioned at the beginning of the proposal that leaves some pods stuck in pending state.
```

 4. Restarts a broker if the broker configuration has changed and cannot be updated dynamically
 5. Does not restart a broker if doing so would take the in-sync replicas count below `min.insync.replicas` for any topic hosted on that broker

#### Combined mode
1. Does not restart a combined node performing log recovery
2. Restarts a combined node if the pod is in stuck state
3. Attempts to connect an admin client to a broker and if it can't connect restarts the broker
4. Restarts a combined node if the controller configuration has changed OR if the broker configuration has changed and cannot be updated dynamically
5. Does not restart a ready combined node if doing so would take the number of caught-up controllers (inc leader) to less than majority of the quorum
6. Does not restart a ready combined node if doing so would take the in-sync replicas count below `min.insync.replicas` for any topic hosted on that broker

#### Unready pod

> Remains the same as Zookeeper mode.

Reviewer 2: ```suggestion
Remains the same as ZooKeeper mode.
```

#### Configuration changes
As implemented in [PR #9125](https://github.com/strimzi/strimzi-kafka-operator/pull/9125) until [KIP 919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum+and+add+Controller+Registration) is implemented:

>  - For broker only nodes, configuration changes are handled in the same way as brokers in the ZooKeeper case. If controller configurations changed, they will be ignored and brokers will not restart.

Reviewer 2: ```suggestion
 - For broker-only nodes, configuration changes are handled in the same way as brokers in the ZooKeeper case. If controller configurations changed, they will be ignored and brokers will not restart.
```
